intermediate_source/custom_function_conv_bn_tutorial.py

```diff
@@ -0,0 +1,344 @@
+        self.fc2 = nn.Linear(128, 10)
+
+    def forward(self, x):
+        if self.fused:
+            x = self.convbn1(x)
+        else:
+            x = self.conv1(x)
+            x = self.bn1(x)
+        x = F.relu(x)
+        if self.fused:
+            x = self.convbn2(x)
+        else:
+            x = self.conv2(x)
+            x = self.bn2(x)
+        x = F.relu(x)
+        x = F.max_pool2d(x, 2)
+        x = F.relu(x)
+        x = x.flatten(1)
+        x = self.fc1(x)
+        x = self.dropout(x)
+        x = F.relu(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+def train(model, device, train_loader, optimizer, epoch):
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        if batch_idx % 2 == 0:
+            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
+                epoch, batch_idx * len(data), len(train_loader.dataset),
+                100. * batch_idx / len(train_loader), loss.item()))
+
+def test(model, device, test_loader):
+    model.eval()
+    test_loss = 0
+    correct = 0
+    # Use inference mode instead of no_grad, for free improved test-time performance
+    with torch.inference_mode():
+        for data, target in test_loader:
+            data, target = data.to(device), target.to(device)
+            output = model(data)
+            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
+            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
+            correct += pred.eq(target.view_as(pred)).sum().item()
+
+    test_loss /= len(test_loader.dataset)
+
+    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
+        test_loss, correct, len(test_loader.dataset),
+        100. * correct / len(test_loader.dataset)))
+
+use_cuda = torch.cuda.is_available()
+
+torch.manual_seed(123456)
+
+device = torch.device("cuda" if use_cuda else "cpu")
+
+train_kwargs = {'batch_size': 4096}
+test_kwargs = {'batch_size': 1024}
+
+if use_cuda:
+    cuda_kwargs = {'num_workers': 1,
+                   'pin_memory': True,
+                   'shuffle': True}
+    train_kwargs.update(cuda_kwargs)
+    test_kwargs.update(cuda_kwargs)
+
+transform = transforms.Compose([
+    transforms.ToTensor(),
+    transforms.Normalize((0.1307,), (0.3081,))
+])
+dataset1 = datasets.MNIST('../data', train=True, download=True,
+                          transform=transform)
+dataset2 = datasets.MNIST('../data', train=False,
+                          transform=transform)
+train_loader = torch.utils.data.DataLoader(dataset1, **train_kwargs)
+test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)
+
+model = Net(fused=True).to(device)
```

Comment:
You could compare the fused model vs the non-fused one by measuring the peak memory usage for the whole training/testing with https://pytorch.org/docs/stable/generated/torch.cuda.max_memory_allocated.html?highlight=max_#torch.cuda.max_memory_allocated

Reply:
This was what I needed!